python_modules/libraries/dagster-components/dagster_components/utils.py

```diff
@@ -143,3 +155,77 @@ def load_module_from_path(module_name, path) -> ModuleType:
     assert spec.loader, "Must have a loader"
     spec.loader.exec_module(module)
     return module
+
+
+# ########################
+# ##### TOML MANIPULATION
+# ########################
+
+
+@contextmanager
+def modify_toml(path: Path) -> Iterator[tomlkit.TOMLDocument]:
+    with open(path) as f:
+        toml = tomlkit.parse(f.read())
+    yield toml
+    with open(path, "w") as f:
+        f.write(tomlkit.dumps(toml))
+
+
+def get_toml_value(doc: tomlkit.TOMLDocument, path: Iterable[str], expected_type: type[T]) -> T:
+    """Given a tomlkit-parsed document/table (`doc`),retrieve the nested value at `path` and ensure
+    it is of type `expected_type`. Returns the value if so, or raises a KeyError / TypeError if not.
+    """
+    current: Any = doc
+    for key in path:
+        # If current is not a table/dict or doesn't have the key, error out
+        if not isinstance(current, dict) or key not in current:
+            raise KeyError(f"Key '{key}' not found in path: {'.'.join(path)}")
+        current = current[key]
+
+    # Finally, ensure the found value is of the expected type
+    if not isinstance(current, expected_type):
+        raise TypeError(
+            f"Expected '{'.'.join(path)}' to be {expected_type.__name__}, "
+            f"but got {type(current).__name__} instead."
+        )
+    return current
+
+
+def set_toml_value(doc: tomlkit.TOMLDocument, path: Iterable[str], value: object) -> None:
+    """Given a tomlkit-parsed document/table (`doc`),set a nested value at `path` to `value`. Raises
+    an error if the leading keys do not already lead to a dictionary.
+    """
+    path_list = list(path)
+    inner_dict = get_toml_value(doc, path_list[:-1], dict)
+    inner_dict[path_list[-1]] = value
```

Comment:
do you anticipate these being used outside tests?

Reply:
moved to test utils